test/clj_puppetdb/http_test.clj

```diff
@@ -52,23 +52,57 @@
         response-data ["node-1" "node-2"]
         response-data-encoded (json/encode response-data)
         response-headers {"x-records" (.toString (count response-data))}
-        fake-get (fn[status] {:status status :body (io/input-stream (.getBytes response-data-encoded)) :headers response-headers})]
+        fake-get (fn [status] {:status status :body (io/input-stream (.getBytes response-data-encoded)) :headers response-headers})]
 
-      (testing "Should have proper response"
-        (with-redefs [http/request-with-client (fn[_ _ _] (future (fake-get 200)))]
-          (let [GET-response (GET client path params)]
-            (is (= (first GET-response) response-data))
-            (is (= (second GET-response) response-headers)))))
+    (testing "Should have proper response"
+      (with-redefs [http/request-with-client (fn [_ _ _] (future (fake-get 200)))]
+        (let [GET-response (GET client path params)]
+          (is (= (first GET-response) response-data))
+          (is (= (second GET-response) response-headers)))))
 
-      (testing "Should throw proper exception"
-        (with-redefs [http/request-with-client (fn[_ _ _] (future (fake-get 400)))]
-          (try
-            (GET client path params)
-            (catch ExceptionInfo ei
-              (let [info (.getData ei)]
-                (is (= (:status info) 400))
-                (is (= (:kind info) :puppetdb-query-error))
-                (is (= (:params info) params))
-                (is (= (:endpoint info) path))
-                (is (= (:host info) host))
-                (is (= (:msg info) response-data-encoded)))))))))
+    (testing "Should throw proper exception"
+      (with-redefs [http/request-with-client (fn [_ _ _] (future (fake-get 400)))]
+        (try
+          (GET client path params)
+          (catch ExceptionInfo ei
+            (let [info (.getData ei)]
+              (is (= (:status info) 400))
+              (is (= (:kind info) :puppetdb-query-error))
+              (is (= (:params info) params))
+              (is (= (:endpoint info) path))
+              (is (= (:host info) host))
+              (is (= (:msg info) response-data-encoded)))))))))
+
+(deftest catching-exceptions-test
+  (testing "Should pass"
+    (is (= (catching-exceptions ((constantly {:body "foobar"})) (assoc-kind {} :something-bad-happened)) {:body "foobar"})))
+  (testing "Should rethrow proper exception on an exception"
+    (try
+      (catching-exceptions (#(throw (NullPointerException.))) (assoc-kind {} :something-bad-happened))
+      (is (not "Should never get to this place!!"))
+      (catch ExceptionInfo ei
+        (let [info (.getData ei)]
+          (is (= (:kind info) :something-bad-happened))
+          (is (instance? NullPointerException (:exception info)))))))
```

Comment:
In a lot of other projects, I've used [slingshot](https://github.com/scgilardi/slingshot) for throwing map exceptions. Slingshot has a `thrown+-with-msg?` assertion type that could possibly be used here, and in the `catching-exceptions` macro we could do `throw+` rather than `throw (ex-info nil ...`. Not something that should be addressed in this PR, just something to think about.